sdk/core/loader/boot.cc

```diff
@@ -460,11 +460,44 @@ namespace
 					    (size == (LA_ABS(__export_mem_heap_end) -
 					              LA_ABS(__export_mem_heap))))
 					{
+						Debug::log("Rounding heap ({}--{}) region",
+						           target,
+						           target + size);
+						// FIXME: This should not be hard coded.
+						static constexpr size_t MantissaBits = 9;
+						static constexpr size_t MantissaMask =
+						  (1 << MantissaBits) - 1;
+						size_t sizeBits = 32 - __builtin_clz(size);
+						size_t sizeMask = (sizeBits <= MantissaBits)
+						                    ? MantissaMask
+						                    : MantissaMask
+						                        << (sizeBits - MantissaBits);
+						Debug::log("Applying mask {} to size", sizeMask);
+						size_t roundedSize = size & sizeMask;
 						Debug::log(
-						  "Assigning the heap ({}--{}) to the allocator",
+						  "Rounding heap size down from {} to {} (rounded up "
+						  "to {})",
+						  size,
+						  roundedSize,
+						  __builtin_cheri_round_representable_length(size));
+						ptraddr_t end         = target + size;
+						ptraddr_t roundedBase = end - roundedSize;
```

Comment:
This is assuming that `end` is suitably aligned.  We should probably assert that `(end & ~sizeMask) == 0`?

That is, I don't think it's worth trying to compensate for the case that it isn't, because it really should be, but all the same, an assertion wouldn't go amiss.